pkg/provider/azure_standard.go

```diff
@@ -1062,7 +1118,80 @@ func (as *availabilitySet) GetNodeNameByIPConfigurationID(ipConfigurationID stri
 	return vmName, strings.ToLower(asName), nil
 }
 
+func (as *availabilitySet) getAvailabilitySetByNodeName(nodeName string, crt azcache.AzureCacheReadType) (*compute.AvailabilitySet, error) {
+	cached, err := as.vmasCache.Get(consts.VMASKey, crt)
+	if err != nil {
+		return nil, err
+	}
+	vmasList := cached.(*sync.Map)
+
+	if vmasList == nil {
+		klog.Warning("Couldn't get all vmas from cache")
+		return nil, nil
+	}
+
+	var result *compute.AvailabilitySet
+	vmasList.Range(func(_, value interface{}) bool {
+		vmasEntry := value.(*availabilitySetEntry)
+		vmas := vmasEntry.vmas
+		if vmas != nil && vmas.AvailabilitySetProperties != nil && vmas.VirtualMachines != nil {
+			for _, vmIDRef := range *vmas.VirtualMachines {
+				if vmIDRef.ID != nil {
+					matches := vmIDRE.FindStringSubmatch(to.String(vmIDRef.ID))
+					if len(matches) != 2 {
+						err = fmt.Errorf("invalid vm ID %s", to.String(vmIDRef.ID))
+						return false
+					}
+
+					vmName := matches[1]
+					if strings.EqualFold(vmName, nodeName) {
+						result = vmas
+						return false
+					}
+				}
+			}
+		}
+
+		return true
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if result == nil {
+		klog.Warningf("failed to find the vmas of node %s", nodeName)
+		return nil, cloudprovider.InstanceNotFound
```

Comment:
let's return nil if vmas is not found (this is because capz would use standalone virtual machines).